File: server/EvalutionAlghoritm/GetIdAudienceForClass.js
import GetRndInteger from "./GetRndInteger.js";

export default function(clas, audiences){
    const id_cathedra = clas.assigned_discipline.specialty.id_cathedra;
    const type_class = clas.id_type_class;
    let detected_audiences = [];
    // Поиск аудиторий у которых закреплены нужные кафедры
    detected_audiences = audiences.filter((aud) => {
        let isFindCathedra = false;
        aud.assigned_audiences.map(au => {
            if(au.id_cathedra === id_cathedra && aud.id_type_class === clas.id_type_class){
                isFindCathedra = true;
            }
        });
        if(isFindCathedra){
            return true;
        }
        else{
            return false;
        }
    })
    // Если в списке всех аудиторий нашлась хоть одна подходящяя с кафедрой и типом, то возвращаем её ид
    if(detected_audiences.length){
        return detected_audiences[GetRndInteger(0, detected_audiences.length-1)].id;
    }
    // Если не найдена ни одна адутитория за кафедрой, то выбираем случайную из всех по типу
    else{
        detected_audiences = audiences.filter(aud => {
            if(aud.id_type_class === clas.id_type_class){
                return true;
            }
        })
        if (!detected_audiences.length){
            return 0;
        }
        return detected_audiences[GetRndInteger(0, detected_audiences.length-1)].id;
    }
}

File: server/EvalationAlgoritmCpp/EvulationAlgoritmCpp/EvulationAlgoritmCpp/GetIdAudienceForClass.cpp
#include "GetIdAudienceForClass.h"
#include "GetRndInteger.h"
#include <iostream>
#include <vector>

using namespace std;

int GetIdAudienceForClass(json clas, json audiences)
{
    vector<int> detected_audiences;
    if (!clas["recommended_audiences"].is_null() && clas["recommended_audiences"].size())
        detected_audiences.push_back(clas["recommended_audiences"][GetRndInteger(0, clas["recommended_audiences"].size())]["id_audience"]);
    else
    {
        int sum_students = 0;
        for (json ag : clas["assigned_groups"])
        {
            sum_students += (int)ag["group"]["number_students"];
        }
        int id_cathedra = clas["assigned_discipline"]["specialty"]["id_cathedra"];

        for (json aud : audiences)
        {
            if (aud["capacity"] >= sum_students && aud["id_type_class"] == clas["id_type_class"])
            {
                for (json au : aud["assigned_audiences"])
                {
                    if (au["id_cathedra"] == id_cathedra)
                    {
                        detected_audiences.push_back(aud["id"]);
                        break;
                    }
                }
            }
        }
        if (!detected_audiences.size())
        {

            for (json aud : audiences)
            {
                if (aud["capacity"] >= sum_students && aud["id_type_class"] == clas["id_type_class"])
                {
                    detected_audiences.push_back(aud["id"]);
                }
            }
            if (!detected_audiences.size())
                return NULL;
        }
    }
    return detected_audiences[GetRndInteger(0, detected_audiences.size() - 1)];
}
